core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java

```diff
@@ -364,12 +360,15 @@ private int updateShardStatuses(EnumSet<ShardStatusChangeType> changes, List<Sha
             return changes;
         }
 
-        private void calcAndSetShardBuffers(int activeShards, String reason) {
-            if (activeShards == 0) {
+        private void calcAndSetShardBuffers(int activeShardCount, String reason) {
```

Comment:
I think things will be simpler if this method is driven by the current shardWasActive map. This means we don't need to count activeShardCount and also not worry about whether availableShards() changed it's mind (see loop at the end of this method)